java/ql/lib/semmle/code/java/security/FragmentInjection.qll

```diff
@@ -0,0 +1,92 @@
+/** Provides classes and predicates to reason about Android Fragment injection vulnerabilities. */
+
+import java
+private import semmle.code.java.dataflow.TaintTracking
+private import semmle.code.java.dataflow.ExternalFlow
+private import semmle.code.java.frameworks.android.Android
+private import semmle.code.java.frameworks.android.Fragment
+private import semmle.code.java.Reflection
+
+/** The method `isValidFragment` of the class `android.preference.PreferenceActivity`. */
+class IsValidFragmentMethod extends Method {
+  IsValidFragmentMethod() {
+    this.getDeclaringType()
+        .getASupertype*()
+        .hasQualifiedName("android.preference", "PreferenceActivity") and
+    this.hasName("isValidFragment")
+  }
+
+  /**
+   * Holds if this method makes the Activity it is declared in vulnerable to Fragment injection,
+   * that is, all code paths in this method return `true` and the Activity is exported.
+   */
+  predicate isUnsafe() {
+    this.getDeclaringType().(AndroidActivity).isExported() and
+    forex(ReturnStmt retStmt, BooleanLiteral bool |
+      retStmt.getEnclosingCallable() = this and
+      // Using taint tracking to handle logical expressions, like
+      // fragmentName.equals("safe") || true
+      TaintTracking::localExprTaint(bool, retStmt.getResult())
```

Comment:
Hmm, this seems FP-ish: for example

```
boolean isValidFragment() {
  boolean result = true;
  if(check1())
    result = false;
  if(result && check2())
    result = false;
  if(result && check3())
    result = false;
  return result;
}
```
Because there's only one return statement, and `true` can get to it, this will be flagged unsafe.

Do you know of cases in the wild where just `forex(ReturnStmt retStmt | retStmt.getResult().(BooleanLiteral).getBooleanValue() = true)` wouldn't suffice?

Reply:
Hmm, I don't think I added this based on any real-world FN cases I spotted. It was more like me trying to prevent those in the first place, but you're right that it also will probably introduce FPs. So I think it makes sense to use your simplified suggestion, and re-visit this in the future if it proves to be insufficient.